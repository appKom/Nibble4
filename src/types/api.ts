export interface UserResponse {
  count: number;
  next: any;
  previous: any;
  results: Array<{
    pk: number;
    first_name: string;
    last_name: string;
    saldo: number;
  }>;
}

export const isUserResponse = (value: any): value is UserResponse => {
  return typeof value.count === "number" || value.results === "object";
};
