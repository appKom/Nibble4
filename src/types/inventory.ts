export type Product = {
  pk: number;
  name: string;
  price: number;
  image: {
    sm: string;
  };
  category: {
    pk: number;
    name: string;
  };
};
