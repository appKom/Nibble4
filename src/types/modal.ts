export enum modalTypes {
  PURCHASE = "PURCHASE",
  COMPLETE = "COMPLETE",
  ERROR = "ERROR",
  DISABLED = "DISABLED",
  INACTIVE = "INACTIVE",
}
