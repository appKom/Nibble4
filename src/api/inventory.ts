import { Product } from "types/inventory";
import { get } from "./index";

const fetchInventory = async (url: string): Promise<Product[]> => {
  const response = await get({ url }).catch(() => Promise.resolve(undefined));
  if (response && response.ok) {
    const json = await response.json();
    return json.sort((a: Product, b: Product) => (a.name > b.name ? 1 : -1));
  }
  return [];
};

export default fetchInventory;
