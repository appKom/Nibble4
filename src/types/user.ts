import { Product } from "./inventory";

export type User = {
  pk: number;
  balance: number;
  first_name: string;
  favourites: Product[];
};
