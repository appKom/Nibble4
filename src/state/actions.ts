import { Product } from "types/inventory";
import { User } from "types/user";
import { modalTypes } from "types/modal";

export enum ActionTypes {
  SET_USER = "SET_USER",
  LOGOUT_USER = "LOGOUT_USER",
  SET_INVENTORY = "SET_INVENTORY",
  SET_FAVOURITES = "SET_FAVOURITES",
  ADD_TO_CART = "ADD_TO_CART",
  REMOVE_FROM_CART = "REMOVE_FROM_CART",
  EMPTY_CART = "EMPTY_CART",
  SET_CATEGORY = "SET_CATEGORY",
  SET_MODAL_STATE = "SET_MODAL_STATE",
}

type SetUserAction = { type: ActionTypes.SET_USER; payload: User };
type LogoutUserAction = { type: ActionTypes.LOGOUT_USER };
type SetInventoryAction = {
  type: ActionTypes.SET_INVENTORY;
  payload: Product[];
};

type SetFavouritesAction = {
  type: ActionTypes.SET_FAVOURITES;
  payload: Product[];
};

type SetCategoryAction = {
  type: ActionTypes.SET_CATEGORY;
  payload: string;
};
type AddToCartAction = {
  type: ActionTypes.ADD_TO_CART;
  payload: number;
};
type RemoveFromCartAction = {
  type: ActionTypes.REMOVE_FROM_CART;
  payload: number;
};

type EmptyCartAction = {
  type: ActionTypes.EMPTY_CART;
};

type SetModalState = {
  type: ActionTypes.SET_MODAL_STATE;
  payload: modalTypes;
};

export type Action =
  | SetUserAction
  | LogoutUserAction
  | SetInventoryAction
  | SetFavouritesAction
  | SetCategoryAction
  | AddToCartAction
  | RemoveFromCartAction
  | EmptyCartAction
  | SetModalState;

export const setUser = (user: User): SetUserAction => ({
  type: ActionTypes.SET_USER,
  payload: user,
});
export const logoutUser = (): LogoutUserAction => ({
  type: ActionTypes.LOGOUT_USER,
});
export const setInventory = (inventory: Product[]): SetInventoryAction => ({
  type: ActionTypes.SET_INVENTORY,
  payload: inventory,
});

export const setFavourites = (favourites: Product[]): SetFavouritesAction => ({
  type: ActionTypes.SET_FAVOURITES,
  payload: favourites,
});

export const addToCart = (id: number): AddToCartAction => ({
  type: ActionTypes.ADD_TO_CART,
  payload: id,
});
export const removeFromCart = (id: number): RemoveFromCartAction => ({
  type: ActionTypes.REMOVE_FROM_CART,
  payload: id,
});
export const emptyCart = (): EmptyCartAction => ({
  type: ActionTypes.EMPTY_CART,
});

export const setCategory = (id: string): SetCategoryAction => ({
  type: ActionTypes.SET_CATEGORY,
  payload: id,
});

export const setModalState = (modalType: modalTypes): SetModalState => ({
  type: ActionTypes.SET_MODAL_STATE,
  payload: modalType,
});
