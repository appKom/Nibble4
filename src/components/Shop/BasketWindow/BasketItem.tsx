import React, { useContext, FC } from "react";
import styled from "styled-components";
import { StateContext } from "state/state";

import { removeFromCart } from "state/actions";
import { deleteItemFromCart } from "state/actions";

import { Product } from "types/inventory";
import { IMAGE_URI } from "../../../api";
import QuantityInput from "atoms/QuantityInput";

type BasketItemProps = {
  id: number;
  quantity: number;
};

const BasketItem: FC<BasketItemProps> = ({ id, quantity }: BasketItemProps) => {
  const { state, dispatch } = useContext(StateContext);

  // const removeItem = () => dispatch(removeFromCart(id));

  const item: Product = state.inventory.find((e) => e.pk === id)!;
  const imageSrc = item.image ? IMAGE_URI(item.image.sm) : "";

  return (
    <Container
      onClick={() => {
        dispatch(removeFromCart(id));
      }}
    >
      <img
        src={
          imageSrc ? imageSrc : `${process.env.PUBLIC_URL}/images/noImage.png`
        }
        alt="Product image"
      />
      <div>
        <b> {item.name} </b>
      </div>
      <RedCrossWrapper>
        <RedCross
          onClick={() => {
            dispatch(deleteItemFromCart(id));
          }}
        >
          <path d="M15.854 12.854c-0-0-0-0-0-0l-4.854-4.854 4.854-4.854c0-0 0-0 0-0 0.052-0.052 0.090-0.113 0.114-0.178 0.066-0.178 0.028-0.386-0.114-0.529l-2.293-2.293c-0.143-0.143-0.351-0.181-0.529-0.114-0.065 0.024-0.126 0.062-0.178 0.114 0 0-0 0-0 0l-4.854 4.854-4.854-4.854c-0-0-0-0-0-0-0.052-0.052-0.113-0.090-0.178-0.114-0.178-0.066-0.386-0.029-0.529 0.114l-2.293 2.293c-0.143 0.143-0.181 0.351-0.114 0.529 0.024 0.065 0.062 0.126 0.114 0.178 0 0 0 0 0 0l4.854 4.854-4.854 4.854c-0 0-0 0-0 0-0.052 0.052-0.090 0.113-0.114 0.178-0.066 0.178-0.029 0.386 0.114 0.529l2.293 2.293c0.143 0.143 0.351 0.181 0.529 0.114 0.065-0.024 0.126-0.062 0.178-0.114 0-0 0-0 0-0l4.854-4.854 4.854 4.854c0 0 0 0 0 0 0.052 0.052 0.113 0.090 0.178 0.114 0.178 0.066 0.386 0.029 0.529-0.114l2.293-2.293c0.143-0.143 0.181-0.351 0.114-0.529-0.024-0.065-0.062-0.126-0.114-0.178z"></path>
        </RedCross>
      </RedCrossWrapper>
      <QuantityInput quantity={quantity} id={id}></QuantityInput>
      <PricetagWrapper>
        <b>{item.price}</b>kr
      </PricetagWrapper>
    </Container>
  );
};

export default BasketItem;

const RedCrossWrapper = styled.div`
  display: flex;
  justify-content: flex-end;
  align-items: end;
`;

const RedCross = styled.svg`
  position: relative; 
  display: block; 
  fill: red;
  height: 1.5em;
  width: 1.5em;
  xmlns: http://www.w3.org/2000/svg;
  background-color: pink;  
`;

const PricetagWrapper = styled.div`
  display: flex;
  justify-content: flex-end;
  align-item: flex-end;
  padding-top: 15px;
`;

const Container = styled.div`
  width: 90%;
  height: 15%;
  margin-left: auto;
  margin-right: auto;
  margin-top: 5px;
  margin-bottom: 15px;
  padding: 10px;
  box-shadow: 2px 2px 7px #888888;
  border-radius: 3px;

  display: grid;
  grid-template-rows: 1fr 1fr;
  grid-template-columns: 1fr 4fr 1fr;

  img {
    width: 75%;
    grid-row: 1 / span 2;
  }
`;
