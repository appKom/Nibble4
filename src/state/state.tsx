import React, { createContext, Dispatch, ReactNode, useReducer } from "react";
import {
  Product,
  CartItem,
  incrementCartItem,
  createCartItem,
  decrementCartItem,
} from "types/inventory";
import { User, OlCoinsUser } from "types/user";
import { modalTypes } from "types/modal";
import { Action, ActionTypes } from "state/actions";

type State = {
  user?: User;
  olCoinsUser?: OlCoinsUser;
  inventory: Product[];
  category: string;
  cart: { [id: number]: CartItem };
  modalState: modalTypes;
  newOlCoins: number;
};

const initialState: State = {
  olCoinsUser: { balance: 100 },
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
  newOlCoins: 0,
};

const getIncrementedCartItem = (id: number, state: State): CartItem => {
  if (state.cart[id]) {
    return incrementCartItem(state.cart[id]);
  }
  return createCartItem(id);
};

const getDecrementedCart = (id: number, state: State) => {
  if (state.cart[id].quantity > 1) {
    return { ...state.cart, [id]: decrementCartItem(state.cart[id]) };
  }
  const removedCart = { ...state.cart };
  delete removedCart[id];
  return { ...removedCart };
};

const updateNewOlCoins = (newValue: number, state: State) => {
  return state.newOlCoins + newValue;
};

const StateReducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.SET_USER:
      return { ...state, user: action.payload };
    case ActionTypes.LOGOUT_USER:
      return {
        ...state,
        user: undefined,
        cart: {},
        category: "Alt",
        modalState: modalTypes.DISABLED,
      };
    case ActionTypes.SET_INVENTORY:
      return { ...state, inventory: action.payload };
    case ActionTypes.SET_CATEGORY:
      return { ...state, category: action.payload };
    case ActionTypes.SET_MODAL_STATE:
      return { ...state, modalState: action.payload };
    case ActionTypes.CHANGE_NEWOLCOINS:
      return {
        ...state,
        newOlCoins: updateNewOlCoins(action.payload, state),
      };
    case ActionTypes.ADD_TO_CART:
      return {
        ...state,
        cart: {
          ...state.cart,
          [action.payload]: getIncrementedCartItem(action.payload, state),
        },
      };
    case ActionTypes.REMOVE_FROM_CART: {
      return {
        ...state,
        cart: getDecrementedCart(action.payload, state),
      };
    }
    case ActionTypes.EMPTY_CART:
      return { ...state, cart: {} };
    default:
      return { ...state };
  }
};

type StateContextProps = {
  state: State;
  dispatch: Dispatch<Action>;
};

const initialContext: StateContextProps = {
  state: initialState,
  dispatch: () => null,
};

export const StateContext = createContext<StateContextProps>(initialContext);

type StateProviderProps = {
  children: ReactNode;
};

export const StateProvider = (props: StateProviderProps): JSX.Element => {
  const [state, dispatch] = useReducer(StateReducer, initialState);
  const { children } = props;

  return (
    <StateContext.Provider value={{ state, dispatch }}>
      {children}
    </StateContext.Provider>
  );
};
