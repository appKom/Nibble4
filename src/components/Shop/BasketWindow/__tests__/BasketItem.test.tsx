import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import BasketItem from "../BasketItem";
import { StateContext, State, StateContextProps } from "state/state";
import { CartItem, createCartItem } from "types/inventory";
import { modalTypes } from "types/modal";
import { Product } from "types/inventory";

const mockProduct: Product = {
  pk: 95,
  name: "Powerking",
  description: "Energidrikk",
  price: 10,
  image: {
    sm: "/media/images/responsive/sm/38688f77-21b6-4936-8ba6-c976538fb110.png",
  },
  category: {
    pk: 3,
    name: "Drikke",
  },
};

const mockCartItem: CartItem = createCartItem(95);

const mockState: State = {
  inventory: [mockProduct],
  category: "Alt",
  cart: { 95: mockCartItem },
  modalState: modalTypes.DISABLED,
};

const fullContext: StateContextProps = {
  state: mockState,
  dispatch: () => null,
};

afterEach(cleanup);

it("Should match snapshot", () => {
  const { asFragment } = render(
    <StateContext.Provider value={fullContext}>
      <BasketItem id={95} quantity={95} />
    </StateContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
