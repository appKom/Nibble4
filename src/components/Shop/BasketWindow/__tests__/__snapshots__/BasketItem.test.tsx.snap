// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ faJJuY"
  >
    <img
      alt="Product image"
      src="https://online.ntnu.no//media/images/responsive/sm/38688f77-21b6-4936-8ba6-c976538fb110.png"
    />
    <span>
      <b>
         Powerking 
      </b>
    </span>
    <span />
    <span>
      Quantity:
      <b>
        95
      </b>
    </span>
    <span>
      <b>
        10
      </b>
      kr
    </span>
  </div>
</DocumentFragment>
`;
